docs/assignment3.md

```diff
@@ -1 +1,418 @@
+| **Intended use**   We will be using this design pattern to ensure our stored data will be accessed and read in a save and correct way by offering a single point of access which we can design and thus control | The main application of this design pattern will be to optimise the use of snippets in our system, creating a lifecycle to minimise disk usage |
+| **Constraints**  | Our solution is also our constraint sinds a big part of our application is now dependend on this pattern to access information. Dependency wise, this is not ideal | * |
+| **Additional remarks**  | * |
+
+We had thoughts about using several design patterns in our application, but decided against for the sake of simplicity, so as not to add patterns that do not solve a real problem. Nevertheless, we will provide an argument for this.
+We used the Singleton pattern on two occasions: for JsonIO and for GistsApi. 
+- JsonIO : It does not make sense to instantiate a json encoder-decoder and pass it to each object when we can have a static one.
+- GistsApi: The same as said about JsonIO. Moreover, GistsApi contains a few parameters, such as
+authorization, that are unique for an instance of the application and can be easily modifiable
+and readable when in a Singleton class.
+
+The patterns we tried / considered using but discarded:
+- Overloading: for a few versions of the app we had an overloaded constructor of the Snippet class that would create the snippet from a gistId. We ended up giving the responsibility to GistsApi to return a snippet when provided with a gistId. 
+The reason behind this is that it shortens the path between the call to the function that should return the snippet and 
+the actual action of getting it from GistsApi. Otherwise we would have to pass maps with snippet arguments from JsonIO to GistsApi to Snippet instead of much more suitable Snippet objects.
+- Composite: We considered extending our app to handle not just single-file snippets but directories of snippets/directories, following the example of Gists. It would be great to implement this design pattern, but, 
+unfortunately, that would be a lot of work to implement this feature.
+- Builder: As it is mentioned in the State Diagrams, there are states of the Snippet when not all of its attributes are/can be defined.
+Then we can gradually add them, this way we would explicitly indicate at what stage of completeness the Snippet is. We thought that the solution would add more complexity
+than the problem creates and decided to use null values with setAttribute(value).
+- Observer: This would come in handy as the application develops: changes to a snippet would be trigger events across the whole application. At this point, changes to a snippet only are trigger events to
+the snippet itself, and are handled by the method onModification().
+- Command: to use with the CliUI to enable advanced behaviour on commands, like queuing and logging, better handling of commands.
+
```

Comment:
-When you state your problem you should specifically note which classes should only have one access point
-I understand your rationale for using Singleton to a certain extent but the wording and description make it hard to do so
-You list some patterns you considered but not really providing enough context as to why you rejected them
-You could have benefitted greatly from using Command and Observer, also Adapter for handling library difficulties or even Decorator in combination with Command to establish different responsibilities for command objects at runtime such as highlighting.
-Overloading isn't a design pattern